### PR TITLE
Moving futures

### PR DESCRIPTION
**Pull request by spall**
There are two main pieces of work here.  First, adding blocking functionality to futures so they will run in the correct continuation.  Second, moving futures from Chez layer to the racket thread layer.  

Blocking:
code was added to core/control.ss that will cause a future to block if it might look past the future's continuation prompt.  This code could probably be improved so fewer things block.  The 'maybe-block' function decides whether to block or not.

Moving futures:
All future code including code for locking and conditions were moved from chez layer to racket thread layer.  All of this code is in a single file to prevent cycles in loading, because locks rely on futures and futures rely on locks.  

In order to implement conditions I modified racket's thread code and created a new sandman.  When a racket thread wants to wait on a condition, the-cw-sandman adds that thread to a list of waiting threads and when the waiting thread is awoken with a signal, the-cw-sandman moves the thread to a list of threads awaiting to be awoken.  
schedule.rkt was changed so the scheduler will not go to sleep if there are threads waiting on conditions, and check-external-events checks the-cw-sandman for any threads waiting to be rescheduled (just like what happens with sleep).  @mflatt, is this the sort of solution you envisioned for this?

I used boxes and 'box-cas!' to implement locks that work with racket threads, futures, and pthreads.  However, I discovered that the racket7 implementation of 'box-cas!' is not actually atomic and cannot be safely used by multiple pthreads.  To fix this, 'box-cas!' now acquires a global mutex.  This is not a long term fix but the code behaves correctly now.

### CODE REVIEW

**Comment by mflatt** [comment]
This generally looks like the right direction, but the scheduler change doesn't look right. By not sleeping when any future is running, aren't you making the main thread busy-wait? The change I had in mind was a sandbman callback that would interrupt a sleep, as opposed to a hook that skips sleeping.

Along those lines, I don't think you want a separate `the-cw-sandman`, but instead extend the base sandman interface. That way, when the sandman is extended in the "io" layer, it can implement its version of sleep with a way to interrupt the sleep. The interrupt step should involve calling `rktio_signal_received_at` to wake up the main thread's sleep (if any is in progress). The `rktio_signal_received_at` function can be called from any OS thread, and its argument indicates which thread's `sleep` to wake up (so that argument will identify the main thread).

**Comment by spall** [comment]
> By not sleeping when any future is running, aren't you making the main thread busy-wait?

It doesn't sleep if there is a thread waiting on a condition; or at least that was my intention.  But I implemented it this way because I didn't know how to wake the scheduler from a sleep.  I can work on implementing this the way you suggest instead. 

>I don't think you want a separate the-cw-sandman, but instead extend the base sandman interface. 

I should add new functions to the sandman struct and then add implementations of those new function
us to the default sandman?

**Comment by mflatt** [comment]
Yes.

There's not an obvious way to interrupt the sleep method that is implemented in "thread/sandman.rkt". But the "io" layer uses a `sleep` implementation that can be interrupted, and so it can implement a new sandman method to interrupt that sleep using `rktio_signal_received_at` – and that's good enough if you don't want to test by running the "thread" layer on its own (by using `make demo` or similar).

File: racket/src/cs/Makefile [hunk 1]
```diff
@@ -5,7 +5,7 @@ SCHEME = scheme
 UNSAFE_COMP = # '(optimize-level 3)'
 
 # Controls whether core is built as unsafe:
-CORE_UNSAFE_COMP = '(optimize-level 3)'
+CORE_UNSAFE_COMP = #'(optimize-level 3)'
```
Comment (mflatt) [inline racket/src/cs/Makefile]: Don't forget to remove this `#`

File: racket/src/cs/Makefile [hunk 2]
```diff
@@ -243,9 +243,7 @@ CORE_SRCS = core/check.ss \
             core/place.ss \
             core/foreign.ss \
             core/future.ss \
-            core/fsemaphore.ss \
-            core/queue.ss \
-            core/future-scheduler.ss
+            core/fsemaphore.ss
```
Comment (mflatt) [inline racket/src/cs/Makefile]: The fsemaphore implementation will also move, right?

File: racket/src/cs/core.sls [hunk 3]
```diff
@@ -566,20 +567,28 @@
...
-	  ;; future scheduler functions
-
-	  future
-	  touch
-	  future?
-	  futures-enabled?
-	  current-future
-	  would-be-future
-	  processor-count
+	  ;; pthread things
```
Comment (mflatt) [inline racket/src/cs/core.sls]: I think we should consistently add "; not exported to Racket" to help others navigate the code, but maybe it makes sense to separate this group and make it a comment for the whole group. (The comment just one the last one here seems misleading.)

File: racket/src/cs/core.sls [hunk 4]
```diff
@@ -641,11 +649,11 @@
...
   (include "core/place.ss")
   (include "core/foreign.ss")
-  (include "core/future.ss")	
-  (include "core/queue.ss")
-  (include "core/future-scheduler.ss")
+  (include "core/lock.ss")
+  (include "core/future.ss")
   
   (set-no-locate-source!)
+  ;; with this below line enabled, errors cause hangs
```
Comment (mflatt) [inline racket/src/cs/core.sls]: Can you make this comment more specific? As-is, it seems to say there's a bug in the implementation, but I think you mean that (other) bugs in the implementation are harder to debug with execptions set up normally.

File: racket/src/cs/core.sls [hunk 5]
```diff
@@ -657,4 +665,8 @@
   (set-impersonator-hash!)
   (set-procedure-impersonator-hash!)
   (set-vector-impersonator-hash!)
-  (set-box-impersonator-hash!))
+  (set-box-impersonator-hash!)
```
Comment (mflatt) [inline racket/src/cs/core.sls]: Revert unnecessary change?

File: racket/src/cs/core/box.ss [hunk 6]
```diff
@@ -2,11 +2,15 @@
...
 
+(define global-lock (make-mutex))
+
 (define (unsafe-box*-cas! b v1 v2)
-  (with-interrupts-disabled
-   (and (eq? v1 (#3%unbox b))
-        (#3%set-box! b v2)
-        #t)))
+  (mutex-acquire  global-lock)
+  (begin0 (with-interrupts-disabled
```
Comment (mflatt) [inline racket/src/cs/core/box.ss]: To make the style more consistent, please use `begin0` only on its own line. (Picky, but being more picky about formatting is one of many things I'm trying out with racket7.) Note also the extra space in the previous line.

File: racket/src/cs/core/control.ss [hunk 7]
```diff
@@ -107,6 +107,12 @@
 ;; `end-uninterrupted` functions bracket such regions dynamically. See
 ;; also "core-engine.ss" and "core-interrupt.ss"
 
+
+;; ---------------------------------------
+
+;; What a futures continuation looks like
```
Comment (mflatt) [inline racket/src/cs/core/control.ss]: ??? A leftover to remove?

File: racket/src/cs/core/control.ss [hunk 8]
```diff
@@ -172,6 +178,27 @@
             #t]
            [else (loop (cdr mc))])))))
 
+(define/who (maybe-block tag)
```
Comment (mflatt) [inline racket/src/cs/core/control.ss]: "block" can mean many things. How about `maybe-future-barricade`, since "barricade" is more specific to the futures design?

File: racket/src/cs/core/control.ss [hunk 9]
```diff
@@ -344,8 +371,11 @@
 
 ;; ----------------------------------------
 
+
+;; This could block because prompt tag could be past future tag
```
Comment (mflatt) [inline racket/src/cs/core/control.ss]: This comment would apply all over and isn't the most important things about `abort-current-continuation`. I think that using a descriptive name like `maybe-future-barricade` makes the comment unnecessary.

File: racket/src/cs/core/control.ss [hunk 10]
```diff
@@ -394,13 +424,15 @@
 (define-record non-composable-continuation full-continuation (tag))
 (define-record escape-continuation continuation (tag))
 
+;; TODO
```
Comment (mflatt) [inline racket/src/cs/core/control.ss]: ?

File: racket/src/cs/core/control.ss [hunk 11]
```diff
@@ -414,12 +446,14 @@
              (extract-metacontinuation 'call-with-current-continuation (strip-impersonator tag) #t)
              tag))))))]))
 
+;; TODO
```
Comment (mflatt) [inline racket/src/cs/core/control.ss]: ?

File: racket/src/cs/core/control.ss [hunk 12]
```diff
@@ -440,6 +474,7 @@
 (define (unsafe-call-with-composable-continuation/no-wind p tag)
   (call-with-composable-continuation* p tag #f))
 
+;; TODO. maybe this is handled elsewhere?
```
Comment (mflatt) [inline racket/src/cs/core/control.ss]: Still TODO? I wouldn't think that anything is needed here.

File: racket/src/cs/core/control.ss [hunk 13]
```diff
@@ -669,7 +704,7 @@
                              (loop rmc))))]))])))
 
 (define (metacontinuation-frame-clear-cache mf)
-  (metacontinuation-frame-merge mf #f))
+  (metacontinuation-frame-merge mf #f)) ;; hmm
```
Comment (mflatt) [inline racket/src/cs/core/control.ss]: ?

File: racket/src/cs/core/control.ss [hunk 14]
```diff
@@ -806,6 +841,8 @@
 (define-record elem+cache (elem cache))
 (define (elem+cache-strip t) (if (elem+cache? t) (elem+cache-elem t) t))
 
+
+;; TODO:
```
Comment (mflatt) [inline racket/src/cs/core/control.ss]: I don't think anything is needed here, since `call-with-immediate-continuation` can't look past the start of a future.

File: racket/src/cs/core/control.ss [hunk 15]
```diff
@@ -1027,6 +1067,7 @@
                                  (vector-set! tmp i ((car wrappers) v))
                                  (key-loop (cdr keys) (cdr wrappers) (add1 i) #t)]))])))]))]))])))))]))
 
+;; TODO
```
Comment (mflatt) [inline racket/src/cs/core/control.ss]: ?

File: racket/src/cs/core/control.ss [hunk 16]
```diff
@@ -1036,12 +1077,14 @@
...
     [(tag)
      (check who continuation-prompt-tag? tag)
+     (maybe-block tag)
      (call/cc
       (lambda (k)
         (make-continuation-mark-set (prune-mark-chain-suffix (strip-impersonator tag) (current-mark-chain))
                                     (cons (continuation->trace k)
                                           (get-metacontinuation-traces (current-metacontinuation))))))]))
 
+;; TODO: could this ever need to block?
```
Comment (mflatt) [inline racket/src/cs/core/control.ss]: Yes, I think this one needs `maybe-future-barricade`.

File: racket/src/cs/core/lock.ss [hunk 17]
```diff
@@ -37,7 +37,8 @@
 
   (define (lock-release lock)
     (when lock
-      (scheduler-lock-release lock)))]
+      (scheduler-lock-release lock)))
+  ]
```
Comment (mflatt) [inline racket/src/cs/core/lock.ss]: Revert unnecessary change?

File: racket/src/cs/core/memory.ss [hunk 18]
```diff
@@ -40,7 +40,8 @@
   (case-lambda
    [() (collect-garbage 'major)]
    [(request)
-    (case request
+    (void)
+    #;(case request
```
Comment (mflatt) [inline racket/src/cs/core/memory.ss:44]: Don't forget to revert this. Does the futures implementation currently need GC to be disabled?

File: racket/src/cs/thread.sls [hunk 19]
```diff
@@ -52,4 +80,8 @@
 
   (set-scheduler-lock-callbacks! (lambda () (1/make-semaphore 1))
                                  1/semaphore-wait
-                                 1/semaphore-post))
+                                 1/semaphore-post)
+
+  (set-future-callbacks! 1/future? 1/current-future
+  			 future-block current-future-prompt))
+
```
Comment (mflatt) [inline racket/src/cs/thread.sls]: Please fix ending (to end in a newline).

File: racket/src/thread/sandman-struct.rkt [hunk 20]
```diff
@@ -30,3 +31,15 @@
                  
   #:prefab)
 
+;; A `sandman` for condition wait.
+(struct cw-sandman sandman
```
Comment (mflatt) [inline racket/src/thread/sandman-struct.rkt]: As discussed: the `cw-sandman` additions need to be part of the main `sandman` API.

File: racket/src/thread/future.rkt [hunk 21]
```diff
@@ -0,0 +1,441 @@
...
+    (begin0
+        (let () expr ...)
+      (lock-release lock))))
+
+(struct future-lock* (box owner))
+
+(define (make-lock)
+  (future-lock* (box 0) (box #f)))
+
+;;  cannot use atomics in locking code because it is used by code that
```
Comment (mflatt) [inline racket/src/thread/future.rkt]: Is it only because you want to check the owner of a lock that locks have to be in "future.rkt"? If so, consider a lock interface where you pass an owner explicitly, and that way locks don't have to be combined here and implemented in multiple places.

File: racket/src/thread/future.rkt [hunk 22]
```diff
@@ -0,0 +1,441 @@
...
+      (if (box-cas! ID id (+ 1 id))
+          id
+          (get-next-id)))))
+
+(define (processor-count)
+  1)
+
+(define futures-enabled? threaded?)
+
+(struct future* (id cond lock prompt
```
Comment (mflatt) [inline racket/src/thread/future.rkt:54]: I think `#:auto` has turned out to be a bad idea. Instead of using `#:auto`, I recommend having a separate `create-future` function that is always used instead of the main constructor (and that fills in all the would-be auto fields with suitable initial values).

File: racket/src/cs/core/memory.ss (hunk 18 above)
Comment (spall) [reply racket/src/cs/core/memory.ss:44]: Yes, there are test cases at call collect-garbage, and chez doesn't allow garbage collection if more than one os thread is running.

Before we decided to move futures from the chez layer to the thread layer I had been working on fixing this.  I had added code that would rendezvous the threads created to run futures, so the main os thread could collect garbage, in the case that collect-garbage was called by an os thread besides the main os thread.

We had discusses before that this would be part of a separate pull request, so I was planning on adding that back after this pull request was merged.  I can revert it and then change it again in the future if you'd prefer.

File: racket/src/cs/thread.sls (hunk 19 above)
Comment (spall) [reply racket/src/cs/thread.sls]: I'm not sure I understand this.  You would like the last empty line to be removed?

Comment (mflatt) [reply racket/src/cs/thread.sls]: Yes, remove the whitespace at the end, especially because it isn't followed by a newline. (By old Unix rules, a text file isn't properly formed unless the last character is a newline. That rule is often ignored now, but it's still a good guideline.)
